tests/TestCase/ORM/MarshallerTest.php

```diff
@@ -558,6 +558,51 @@ public function testOneBelongsToManyJoinDataAssociatedWithIds()
     }
 
     /**
+     * Test belongsToMany association with mixed data array
+     *
+     * @return void
+     */
+    public function testBelongsToManyWithMixedData()
+    {
+
+        $data = [
+            'title' => 'My title',
+            'body' => 'My content',
+            'author_id' => 1,
+            'tags' => [
+                [
+                    'name' => 'tag4'
+                ],
+                [
+                    'name' => 'tag5'
+                ],
+                [
+                    'id' => 1
+                ]
+            ]
+        ];
+
+        $articles = TableRegistry::get('Articles');
+        $articles->belongsToMany('Tags');
+
+        $tags = TableRegistry::get('Tags');
+
+        $article = $articles->newEntity($data, [
+            'associated' => [
+                'Tags'
+            ]
+        ]);
+
+        $this->assertEquals($article->tags[2], $tags->get(1));
+
+        $tag_count = $tags->find()->count();
+        $articles->save($article);
+
+        $this->assertEquals($tag_count + 2, $tags->find()->count());
```

Comment:
You should also assert that indexes 0, 1 marshalled correctly and are marked as isNew().